editor/src/components/canvas/canvas-wrapper-component.tsx

```diff
@@ -93,11 +119,296 @@ export const CanvasWrapperComponent = React.memo(() => {
+      width:
+        Math.max(selectionAreaStart.x, mousePoint.x) - Math.min(selectionAreaStart.x, mousePoint.x),
+      height:
+        Math.max(selectionAreaStart.y, mousePoint.y) - Math.min(selectionAreaStart.y, mousePoint.y),
+    })
+  }, [mousePoint, selectionAreaStart])
+
+  const canSelectArea = React.useMemo(() => {
+    if (selectionArea != null || isSelectModeWithArea(mode)) {
+      return true
+    }
+    if (mode.type !== 'select') {
+      return false
+    }
+    if (!mouseIsOverStoryboardOrEmptyScene) {
+      return false
+    }
+    return true
+  }, [mode, mouseIsOverStoryboardOrEmptyScene, selectionArea])
+
+  const canvasSelectionArea = React.useMemo(() => {
+    const refRect = ref.current?.getBoundingClientRect() ?? null
+    if (selectionArea == null || refRect == null) {
+      return null
+    }
+
+    return canvasRectangle({
+      x: selectionArea.x - refRect.x - actualNavigatorWidth,
+      y: selectionArea.y - refRect.y,
+      width: selectionArea.width,
+      height: selectionArea.height,
+    })
+  }, [selectionArea, actualNavigatorWidth])
+
+  function onMouseMove(e: React.MouseEvent) {
+    setMousePoint(canvasPoint({ x: e.clientX, y: e.clientY }))
+  }
+
+  const isValidMouseEventForSelectionArea = React.useCallback(
+    (e: React.MouseEvent): boolean => {
+      return (
+        mousePoint != null && e.button === 0 && !(e.shiftKey || e.metaKey || e.ctrlKey || e.altKey)
+      )
+    },
+    [mousePoint],
+  )
+
+  const onMouseDown = React.useCallback(
+    (e: React.MouseEvent) => {
+      if (isValidMouseEventForSelectionArea(e)) {
+        if (canSelectArea && selectionAreaStart == null) {
+          setSelectionAreaStart(mousePoint)
+          setElementsUnderSelectionArea([])
+          dispatch([switchEditorMode(EditorModes.selectMode(null, true)), clearSelection()])
+        }
+      }
+    },
+    [canSelectArea, mousePoint, dispatch, selectionAreaStart, isValidMouseEventForSelectionArea],
+  )
+
+  const clearAndGetActions = React.useCallback((): EditorAction[] => {
+    if (mode.type !== 'select') {
+      return []
+    }
+    return [
+      switchEditorMode(EditorModes.selectMode()),
+      clearHoveredViews(),
+      clearHighlightedViews(),
+    ]
+  }, [mode.type])
+
+  const onMouseUp = React.useCallback(
+    (e: React.MouseEvent) => {
+      setSelectionAreaStart(null)
+      let actions: EditorAction[] = clearAndGetActions()
+      if (
+        selectionAreaStart != null &&
+        highlightedViews.length > 0 &&
+        isValidMouseEventForSelectionArea(e)
+      ) {
+        actions.unshift(selectComponents(highlightedViews, false))
+      }
+      dispatch(actions)
+    },
+    [
+      dispatch,
+      selectionAreaStart,
+      highlightedViews,
+      clearAndGetActions,
+      isValidMouseEventForSelectionArea,
+    ],
+  )
+
+  React.useEffect(() => {
+    if (selectionArea == null || mousePoint == null || selectionAreaStart == null) {
```

Comment:
Can this be moved into a standalone function? This is a lot of logic, and probably it could be broken up even further. Also, if the logic here only run during an area select session, I think a lot of things can be precomputed here